### PR TITLE
[59707] Java SDK support for metadata for events

### PR DESCRIPTION
**Pull request by mrashed-dev**
# Description
The Nylas Calendar API has a new beta feature of adding a new [metadata field to calendar events](https://www.nylas.com/blog/event-metadata). This PR makes the feature available via the Nylas Java SDK. 

# Usage
To create a new event with event metadata, you can create a `Map<String, String>` with a mapping of key-value pairs for the metadata:
```
Event event = new Event(calendarId, new Timespan(startTime, endTime));
....
Map<String, String> metadata = new HashMap<>();
metadata.put("event_category", "gathering");
event.setMetadata(metadata);
```

To query events based on metadata, you can filter on three different parameters:
- `metadata_key` (string or array) to filter based on the keys within the metadata object
- `metadata_value` (string or array) to filter based on the value within the metadata object
- `metadata_pair` (pair of strings; a key and a value) to filter based on the key-value pairs within the metadata object

You can invoke them as such and even chain them:
```
EventQuery query = new EventQuery()
                .calendarId(calendarId)
                .metadataKey("visitors", "parking")
                .metadataValue("garden")
                .metadataPair("event_category", "gathering");
```

# License
I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: src/main/java/com/nylas/Event.java
Comment (ddlatham) [inline src/main/java/com/nylas/Event.java:26]: Since metadata appears to be limited to string values, I think it's best to model it is as a `Map<String, String>` instead,

https://developer.nylas.com/docs/connectivity/calendar/event-metadata/#event-metadata-object

**Review by philrenaud — APPROVED** [review]

File: src/main/java/com/nylas/Event.java
Comment (mrashed-dev) [reply src/main/java/com/nylas/Event.java]: Good call! Thanks for the link!

File: src/main/java/com/nylas/Event.java
Comment (ddlatham) [inline src/main/java/com/nylas/Event.java:161]: turns out that the server gives a 500 error when doing a PUT with an empty metadata map.  need to get that fixed or add a special case here to omit empty maps

https://app.clubhouse.io/nylas/story/59971/put-event-with-empty-metadata-results-in-500-error

Comment (ddlatham) [inline src/main/java/com/nylas/Event.java:161]: please do not merge until we decide on the plan for this issue

File: src/main/java/com/nylas/EventQuery.java
Comment (ddlatham) [inline src/main/java/com/nylas/EventQuery.java:148]: Since we ended up working concurrently on this support, I have another implementation that I would prefer if you don't mind.  It looks like this:

Fields:
```	private List<String> metadataKeys;
	private List<String> metadataValues;
	private List<String> metadataPairs;
```

addParameters:
```
		if (metadataKeys != null) {
			for (String key : metadataKeys) {
				url.addQueryParameter("metadata_key", key);
			}
		}
		if (metadataValues != null) {
			for (String value : metadataValues) {
				url.addQueryParameter("metadata_value", value);
			}
		}
		if (metadataPairs != null) {
			for (String pair : metadataPairs) {
				url.addQueryParameter("metadata_pair", pair);
			}
		}
```

fluent filter methods:
```
	/**
	 * Add a required metadata key to the query.
	 * Returned events must include a metadata property with this key.
	 * This method can be invoked multiple times to require the event to have multiple
	 * metadata keys.  Each time this is called it adds another required key rather than
	 * replacing previous required keys.
	 */
	public EventQuery metadataKey(String metadataKey) {
		if (this.metadataKeys == null) {
			this.metadataKeys = new ArrayList<>();
		}
		this.metadataKeys.add(metadataKey);
		return this;
	}	
	
	/**
	 * Add a required metadata value to the query.
	 * Returned events must include a metadata property with this value.
	 * This method can be invoked multiple times to require the event to have multiple
	 * metadata values.  Each time this is called it adds another required value rather than
	 * replacing previous required values.
	 */
	public EventQuery metadataValue(String metadataValue) {
		if (this.metadataValues == null) {
			this.metadataValues = new ArrayList<>();
		}
		this.metadataValues.add(metadataValue);
		return this;
	}
	
	/**
	 * Add a required metadata key/value pair to the query.
	 * Returned events must include a metadata property with this key and value.
	 * This method can be invoked multiple times to require the event to have multiple
	 * metadata key value pairs.  Each time this is called it adds another required key value pair rather than
	 * replacing previous required key value pairs.
	 */
	public EventQuery metadataKeyValue(String key, String value) {
		if (this.metadataPairs == null) {
			this.metadataPairs = new ArrayList<>();
		}
		this.metadataPairs.add(key + ":" + value);
		return this;
	}
```

I prefer this style of fluent methods as it allows adding a single filter at a time rather than replacing all of them.

**Review by ddlatham — CHANGES_REQUESTED** [review]
Left a couple comments.  I also have a few question still for @rajeev2gupta in this story that I'd like to pin down before merging this in.  May require some additional documentation:
https://app.clubhouse.io/nylas/story/57542/java-sdk-support-for-metadata-for-events

File: src/main/java/com/nylas/Event.java
Comment (mrashed-dev) [reply src/main/java/com/nylas/Event.java:161]: Good idea. Holding off.

File: src/main/java/com/nylas/EventQuery.java
Comment (mrashed-dev) [reply src/main/java/com/nylas/EventQuery.java:179]: Changes implemented!

File: src/main/java/com/nylas/Event.java
Comment (mrashed-dev) [reply src/main/java/com/nylas/Event.java:161]: The issue has been fixed and deployed in Prod. Example code reflects the current working state.

File: CHANGELOG.md
Comment (ddlatham) [inline CHANGELOG.md:12]: Looks like a merge conflict didn't resolve quite right.  Since 1.7.0 was released, the Calendar.isPrimary line is already included there below.

Also, I'd suggest keeping a simpler one line for event metadata.  Something like:
`- Event metadata support added`

File: src/examples/java/com/nylas/examples/other/EventsExample.java
Comment (ddlatham) [inline src/examples/java/com/nylas/examples/other/EventsExample.java:103]: alternatively, `created.setMetadata(Maps.of("event_category", "gathering"));`
(using `com.nylas.Maps`)
but this is fine too.

File: CHANGELOG.md
Comment (mrashed-dev) [reply CHANGELOG.md:12]: Good catch, fixed the changelog!

File: src/examples/java/com/nylas/examples/other/EventsExample.java
Comment (mrashed-dev) [reply src/examples/java/com/nylas/examples/other/EventsExample.java:103]: For now we will keep it this way as `com.nylas.Maps` is a private class.

File: src/main/java/com/nylas/EventQuery.java
Comment (ddlatham) [inline src/main/java/com/nylas/EventQuery.java:141]: Rajeev supplied answers that multiple `metadata` filters (of the same type or different types) are ORed together, unlike other filters.  We need to update the javadoc on the three methods here to reflect that.  Here's a suggestion:

```
/**
 * Return events with metadata containing a property having the given key.
 * 
 * If multiple instances of metadata methods are invoked
 * (any combination of calls to metadataKey, metadataValue, or metadataPair),
 * then this query will return events which match ANY one of them.  
 */
```

Comment (mrashed-dev) [reply src/main/java/com/nylas/EventQuery.java]: Awesome, added in now.

**Review by ddlatham — APPROVED** [review]
Looks good to merge!

**Pull request merged**
